File: backend/schemas/livro.py
from datetime import datetime

from pydantic import BaseModel, Field, ConfigDict


# =====================================================
# LIVRO BASE
# =====================================================

class LivroBase(BaseModel):

    titulo: str = Field(
        min_length=1,
        max_length=255
    )

    subtitulo: str | None = Field(
        default=None,
        max_length=255
    )

    isbn: str | None = Field(
        default=None,
        max_length=20
    )

    sinopse: str | None = None

    ano_publicacao: int | None = Field(
        default=None,
        gt=0
    )

    numero_paginas: int | None = Field(
        default=None,
        gt=0
    )

    idioma: str | None = Field(
        default=None,
        max_length=50
    )

    capa: str | None = Field(
        default=None,
        max_length=500
    )

    editora_id: int


# =====================================================
# CRIAÇÃO DE LIVRO
# =====================================================

class LivroCreate(LivroBase):

    pass


# =====================================================
# ATUALIZAÇÃO DE LIVRO
# =====================================================

class LivroUpdate(BaseModel):

    titulo: str

    subtitulo: str | None = None

    isbn: str | None = None

    sinopse: str | None = None

    ano_publicacao: int | None = None

    numero_paginas: int | None = None

    idioma: str | None = None

    capa: str | None = None

    editora_id: int


# =====================================================
# RESPOSTA COMPLETA DO LIVRO
# =====================================================

class LivroResponse(BaseModel):

    id: int

    titulo: str

    subtitulo: str | None = None

    isbn: str | None = None

    sinopse: str | None = None

    ano_publicacao: int | None = None

    numero_paginas: int | None = None

    idioma: str | None = None

    capa: str | None = None

    editora_id: int

    data_cadastro: datetime

    model_config = ConfigDict(
        from_attributes=True
    )


# =====================================================
# AUTOR NA LISTAGEM DE LIVROS
# =====================================================

class AutorLivroResponse(BaseModel):

    id: int

    nome: str

    model_config = ConfigDict(
        from_attributes=True
    )


# =====================================================
# GÊNERO NA LISTAGEM DE LIVROS
# =====================================================

class GeneroLivroResponse(BaseModel):

    id: int

    nome: str

    model_config = ConfigDict(
        from_attributes=True
    )


# =====================================================
# RESPOSTA DA LISTAGEM DE LIVROS
# =====================================================

class LivroListaResponse(BaseModel):

    id: int

    titulo: str

    ano_publicacao: int | None

    idioma: str | None

    numero_paginas: int | None

    editora: str | None

    autores: list[AutorLivroResponse]

    generos: list[GeneroLivroResponse]

    media_avaliacoes: float | None

    quantidade_avaliacoes: int

    model_config = ConfigDict(
        from_attributes=True
    )

File: frontend/js/livro.js
// =====================================================
// PÁGINA DE DETALHES DO LIVRO
// =====================================================


// -----------------------------------------------------
// Ao carregar a página
// -----------------------------------------------------

document.addEventListener('DOMContentLoaded', () => {

    carregarLivro();

});


// -----------------------------------------------------
// Carrega o livro através do ID da URL
// -----------------------------------------------------

async function carregarLivro() {

    const params = new URLSearchParams(window.location.search);

    const livroId = params.get('id');


    // Verifica se existe um ID na URL

    if (!livroId) {

        mostrarErro('Livro não informado.');

        return;

    }


    try {

        const livro = await getLivro(livroId);

        preencherLivro(livro);

        await carregarAutores(livroId);

        await carregarGeneros(livroId);

    } catch (error) {

        console.error('Erro ao carregar livro:', error);

        mostrarErro('Não foi possível carregar as informações do livro.');

    }

}


// -----------------------------------------------------
// Preenche as informações principais
// -----------------------------------------------------

function preencherLivro(livro) {


    // Título

    document.getElementById('book-title').innerText =
        livro.titulo || 'Título não informado';


    // Subtítulo

    const subtitleElement =
        document.getElementById('book-subtitle');

    if (livro.subtitulo) {

        subtitleElement.innerText = livro.subtitulo;

    } else {

        subtitleElement.innerText = '';

    }


    // Ano

    document.getElementById('book-year').innerText =
        livro.ano_publicacao || 'Não informado';


    // Páginas

    document.getElementById('book-pages').innerText =
        livro.numero_paginas
            ? `${livro.numero_paginas} páginas`
            : 'Não informado';


    // Idioma

    document.getElementById('book-language').innerText =
        livro.idioma || 'Não informado';


    // ISBN

    document.getElementById('book-isbn').innerText =
        livro.isbn || 'Não informado';


    // Sinopse

    const synopsisElement =
        document.getElementById('book-synopsis');

    if (livro.sinopse) {

        synopsisElement.innerHTML =
            `<p>${livro.sinopse}</p>`;

    } else {

        synopsisElement.innerHTML =
            `<p class="text-muted">
                Sinopse não informada.
            </p>`;

    }


    // Capa

    const coverContainer =
        document.getElementById('book-cover-container');


    if (livro.capa) {

        coverContainer.innerHTML = `
            <img
                src="${livro.capa}"
                alt="Capa de ${livro.titulo}"
                class="book-detail-cover"
            >
        `;

    } else {

        coverContainer.innerHTML = `
            <div class="book-cover-placeholder">
                📚
            </div>
        `;

    }


    // Data de cadastro

    if (livro.data_cadastro) {

        const data =
            new Date(livro.data_cadastro);

        document.getElementById(
            'book-registration-date'
        ).innerText =
            data.toLocaleDateString('pt-BR');

    }


    // Editora

    // Por enquanto o GET /livros/{id} retorna apenas editora_id.
    document.getElementById('book-publisher').innerText =
        livro.editora_id
            ? `Editora #${livro.editora_id}`
            : 'Não informado';

}


// -----------------------------------------------------
// Avaliações
// -----------------------------------------------------

const ratingElement =
    document.getElementById('book-rating');

ratingElement.innerHTML = `
    <span class="text-muted">
        <i class="fa-regular fa-star me-1"></i>
        Avaliações serão carregadas em breve.
    </span>
`;




// -----------------------------------------------------
// Carrega autores
// -----------------------------------------------------

async function carregarAutores(livroId) {

    const container =
        document.getElementById('book-authors');


    try {

        const autores =
            await getAutoresDoLivro(livroId);


        if (!autores || autores.length === 0) {

            container.innerHTML = `
                <p class="text-muted">
                    Autor não informado.
                </p>
            `;

            return;

        }


        container.innerHTML = autores.map(autor => `

            <div class="author-card">

                <div class="author-icon">

                    <i class="fa-solid fa-user"></i>

                </div>

                <div>

                    <h5 class="mb-1">
                        ${autor.nome}
                    </h5>

                </div>

            </div>

        `).join('');


    } catch (error) {

        console.error(
            'Erro ao carregar autores:',
            error
        );

        container.innerHTML = `
            <p class="text-muted">
                Não foi possível carregar os autores.
            </p>
        `;

    }

}


// -----------------------------------------------------
// Carrega gêneros
// -----------------------------------------------------

async function carregarGeneros(livroId) {

    const container =
        document.getElementById('book-genre');


    try {

        const generos =
            await getGenerosDoLivro(livroId);


        if (!generos || generos.length === 0) {

            container.innerText =
                'Gênero não informado.';

            return;

        }


        container.innerText =
            generos
                .map(genero => genero.nome)
                .join(' • ');


    } catch (error) {

        console.error(
            'Erro ao carregar gêneros:',
            error
        );

        container.innerText =
            'Não foi possível carregar o gênero.';

    }

}


// -----------------------------------------------------
// Exibe erro na página
// -----------------------------------------------------

function mostrarErro(mensagem) {

    document.getElementById('book-title').innerText =
        'Não foi possível carregar o livro.';

    document.getElementById('book-synopsis').innerHTML = `
        <p class="text-danger">
            ${mensagem}
        </p>
    `;

}

